cmd/snap/cmd_debug_state.go

```diff
@@ -293,6 +301,165 @@ func (c *cmdDebugState) showIsSeeded(st *state.State) error {
+	defer st.Unlock()
+
+	p := strings.FieldsFunc(connArg, func(r rune) bool {
+		return r == ' ' || r == ','
+	})
+
+	var plugMatch, slotMatch SnapAndName
+	if err := plugMatch.UnmarshalFlag(p[0]); err != nil {
+		return err
+	}
+
+	if len(p) > 1 {
+		if err := slotMatch.UnmarshalFlag(p[1]); err != nil {
+			return err
+		}
+	}
+
+	var conns map[string]*connState
+	if err := st.Get("conns", &conns); err != nil && !errors.Is(err, state.ErrNoState) {
+		return err
+	}
+
+	// sort by connection ID
+	connIDs := make([]string, 0, len(conns))
+	for connID := range conns {
+		connIDs = append(connIDs, connID)
+	}
+	sort.Strings(connIDs)
+
+	for _, connID := range connIDs {
+		connRef, err := interfaces.ParseConnRef(connID)
+		if err != nil {
+			return err
+		}
+
+		refMatch := func(x SnapAndName, y interface{ String() string }) bool {
+			parts := strings.Split(y.String(), ":")
+			return len(parts) == 2 && x.Snap == parts[0] && x.Name == parts[1]
+		}
+		plug, slot := connRef.PlugRef, connRef.SlotRef
+
+		if slotMatch.Name != "" && slotMatch.Snap != "" && plugMatch.Snap != "" && plugMatch.Name != "" {
+			// command invoked with 'snap:plug,snap:slot'
+			if !refMatch(plugMatch, plug) || !refMatch(slotMatch, slot) {
+				// should match the connection exactly
+				continue
+			}
+		} else if plugMatch.Snap != "" && plugMatch.Name != "" && slotMatch.Snap == "" && slotMatch.Name == "" {
+			// command invoked with 'snap:plug-or-slot'
+			if !refMatch(plugMatch, plug) && !refMatch(plugMatch, slot) {
+				// should match either the connection's slot or plug
+				continue
+			}
+		} else if plugMatch.Snap != "" && plugMatch.Name == "" && slotMatch.Snap == "" && slotMatch.Name == "" {
+			// command invoked with 'snap' only
+			if plugMatch.Snap != slot.Snap && plugMatch.Snap != plug.Snap {
+				// should match one of the snap names
+				continue
+			}
+		} else {
+			return fmt.Errorf("invalid command with connection args: %s", connArg)
+		}
+
+		conn := conns[connID]
+
+		// the output of 'debug connection' is yaml
+		fmt.Fprintf(Stdout, "id: %s\n", connID)
+		out, err := yaml.Marshal(conn)
+		if err != nil {
+			return err
+		}
+		fmt.Fprintf(Stdout, "%s\n", out)
+	}
+	return nil
+}
+
+func (c *cmdDebugState) showConnections(st *state.State) error {
+	st.Lock()
+	defer st.Unlock()
+
+	var conns map[string]*connState
+	if err := st.Get("conns", &conns); err != nil && !errors.Is(err, state.ErrNoState) {
+		return err
+	}
+
+	all := make([]*connectionInfo, 0, len(conns))
+	for connID, conn := range conns {
+		p := strings.Split(connID, " ")
+		if len(p) != 2 {
+			return fmt.Errorf("cannot parse connection ID %q", connID)
+		}
+		plug := strings.Split(p[0], ":")
+		slot := strings.Split(p[1], ":")
+
+		c := &connectionInfo{
+			PlugSnap:  plug[0],
+			PlugName:  plug[1],
+			SlotSnap:  slot[0],
+			SlotName:  slot[1],
+			connState: *conn,
+		}
+		all = append(all, c)
+	}
+
+	sort.Sort(byPlug(all))
+
+	w := tabwriter.NewWriter(Stdout, 5, 3, 2, ' ', 0)
```

Comment:
This looks similar to `cmd_connections.go` - I wonder if there is anything we could extract as a common helper here?

Reply:
I could make a NewDefaultWriter but picking these values as default feels somewhat arbitrary without a reasoning behind it. I grepped for this but didn't find a comment on why these particular values were chosen

Reply:
Sorry, I meant most of the output code looks similar not just this one line. But it's fine I think, I did not look very deep was mostly an idle wondering.

Reply:
(And it's fine as is to be clear)